### DISCUSSION TITLE
Upload image but got an error "Invalid key"

### DISCUSSION THREAD

**Discussion by hallee9000** [supabase/supabase#2480] [Questions] [answered]
Hi there, I've been trying to upload an image but got an error.

![image](https://user-images.githubusercontent.com/12868055/126502913-0c31a690-c924-44af-851a-60a0d736ced2.png)

My code is like:
```ts
export async function uploadFile (file: any) {
  const { data, error } = await supabase
    .storage
    .from('project')
    .upload(uuid() + '-' + encodeURIComponent(file.name), file)
  console.log(data, error)
  return data
}
```

The file path is like `3d1f59d4-f66a-4b11-b5e7-b46d4c00eb06-%E4%B8%8B%E8%BD%BD.gif`. The file name contains Chinese characters so I encoded it.

BTW `upload` method will return file key(path) with bucket id but `getPublicUrl` method doesn't need bucket id. Does this mean I have to remove bucket id manually from file key before saving it in the database?

**Comment by dshukertjr** [comment, marked as answer]
Hi @leadream !

It seems like you cannot have `%` within your file path. 
I checked inside [storage-api repo](https://github.com/supabase/storage-api), and there was a validation like this for the file path:
```typescript
export function isValidKey(key: string): boolean {
  // only allow s3 safe characters and characters which require special handling for now
  // https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
  return /^(\w|\/|!|-|\.|\*|'|\(|\)| |&|\$|@|=|;|:|\+|,|\?)*$/.test(key)
}
```

>BTW upload method will return file key(path) with bucket id but getPublicUrl method doesn't need bucket id. Does this mean I have to remove bucket id manually from file key before saving it in the database?

Actually, `getPublicUrl` does require a bucket ID. In the following example `project` is the bucket ID. There is no manual work of removing anything when using `getPublicUrl`. You should be able to use it like this:
```typescript
  const filePath = uuid() + '-' + file.name
  const { data, error } = await supabase.storage.from('project').upload( filePath, file )
  const { publicURL } = supabase.storage.from('project').getPublicUrl(filePath)
```

Hopefully this answer clarified your question, but if not, please don't hesitate to ask any followup questions that you may have!

**Reply by hallee9000** [reply]
Thanks @dshukertjr that's very helpful. I'll try it later.

**Reply by roker15** [reply]
> Hi @leadream !
> 
> It seems like you cannot have `%` within your file path. I checked inside [storage-api repo](https://github.com/supabase/storage-api), and there was a validation like this for the file path:
> 
> ```ts
> export function isValidKey(key: string): boolean {
>   // only allow s3 safe characters and characters which require special handling for now
>   // https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
>   return /^(\w|\/|!|-|\.|\*|'|\(|\)| |&|\$|@|=|;|:|\+|,|\?)*$/.test(key)
> }
> ```
> 
> > BTW upload method will return file key(path) with bucket id but getPublicUrl method doesn't need bucket id. Does this mean I have to remove bucket id manually from file key before saving it in the database?
> 
> Actually, `getPublicUrl` does require a bucket ID. In the following example `project` is the bucket ID. There is no manual work of removing anything when using `getPublicUrl`. You should be able to use it like this:
> 
> ```ts
>   const filePath = uuid() + '-' + file.name
>   const { data, error } = await supabase.storage.from('project').upload( filePath, file )
>   const { publicURL } = supabase.storage.from('project').getPublicUrl(filePath)
> ```
> 
> Hopefully this answer clarified your question, but if not, please don't hesitate to ask any followup questions that you may have!

Then what is the practical use of data.key? can u tell me it is used nowhere, if store this in database what i can do with this in future? i think data.key is useless.

**Comment by grukillerkams** [comment]
Hi There, I am facing the same issue with this line:

`await supabase.storage.from("audios").upload("", data.file);`

this is my file body:

```javascript
File {name: 'dsas.m4a', lastModified: 1726358253544, lastModifiedDate: Sun Sep 15 2024 00:57:33 GMT+0100 (West Africa Standard Time), webkitRelativePath: '', size: 40446, …}lastModified: 1726358253544lastModifiedDate: Sun Sep 15 2024 00:57:33 GMT+0100 (West Africa Standard Time) {}name: "dsas.m4a"size: 40446type: "audio/x-m4a"webkitRelativePath: ""[[Prototype]]: File
```

this is the error response object:

```javascript
{statusCode: '400', error: 'InvalidKey', message: 'Invalid key: '}error: "InvalidKey"message: "Invalid key: "statusCode: "400"[[Prototype]]: Object
```
